2.md

```diff
@@ -0,0 +1,278 @@
+
+The first argument to the `Module` constructor is a source.
+If that source is an object that does not have a [[Module Source]] internal
+slot, we treat this object as a protocol for loading, binding, linking,
+initializing, and executing the new `Module` instance.
+
+## Interface
+
+```ts
+type VirtualModuleSource = {
+  // Indicates the import and export bindings the module has
+  // between its module environment record, module namespace exotic object,
+  // and its dependencies.
+  bindings?: Array<Binding>,
+
+  // Executes the module if it is imported.
+  // execute may return a promise, indicating that the module uses
+  // the equivalent of top-level-await.
+  execute?: (namespace: ModuleImportsNamespace, {
+    import?: (importSpecifier: string) => Promise<ModuleExportsNamespace>,
+    importMeta?: Object
+  }) => void,
+
+  // Indicates that execute needs to receive a dynamic import function
+  // bound to a Module instance.
+  needsImport?: boolean,
+
+  // Indicates that initialize needs to receive an importMeta.
+  needsImportMeta?: boolean,
+};
+```
+
+Bindings must be one of the shapes proposed in [module source static
+analysis](./1.md), such that for each binding gets linked in a virtual module.
+
+The `Module` constructor from [Module and ModuleSource](./0.md) extends to
+accept virtual module sources instead of `ModuleSource` and reflects whatever
+`source` its given as the `source` property of the returned instance.
+
+## Examples
+
+### JSON
+
+This protocol allows user code to create new module source constructors.
+For example, for a host that does not support JSON, this could be
+accomplished with a small virtual module source.
+
+```js
+class JsonModuleSource {
+  bindings = { export: 'default' };
+  constructor(text) {
+    this.value = JSON.parse(text);
+  };
+  execute(imports) {
+    imports.default = this.value;
+  };
+}
+
+const source = new JsonModuleSource({ meaning: 42 });
+const module = new Module(source);
+const { default: { meaning } } = await import(module);
+```
+
+Asset modules of various kinds would largely follow this pattern.
+
+### WASM
+
+On a host that does not provide support for WebAssembly, a virtual module
+source would suffice.
+
+```js
+class WasmModuleSource {
+  constructor(buffer) {
+    const module = new WebAssembly.Module(buffer);
+    this.#imports = WebAssembly.Module.imports(module);
+    this.#exports = WebAssembly.Module.exports(module);
+    this.bindings = [
+       ...this.#imports.map(({ module, name }) =>
+         ({ import: name, from: module })),
+       ...this.#exports.map(({ name }) =>
+         ({ export: name })),
+    ];
+  };
+  async execute(namespace) {
+    const importObject = {};
+    for (const { module, name, kind } of this.#imports) {
+      importObject[name] = namespace[name];
+    }
+    const instance = await WebAssembly.instantiate(module, importObject);
+    for (const { name } of this.#exports) {
+      namespace[name] = instance[name];
+    }
+  };
```

Comment:
Is `namespace` both for receiving the imports, and for exporting them? What if I want to export something with the same name as an imported value?

Reply:
Oh that's probably why we have `as` for imports.

Reply:
Avoiding such collisions is the purview of bindings, where `as` clauses can resolve conflicts. The namespace object reflects the bound subset of the names an ESM will have in its global contour.